### PR TITLE
[BEAM-6772] Change Select semantics to match what a user expects

### PR DESCRIPTION
**Pull request by reuvenlax**
Details in the JIRA: The current Select transform provides confusing and less-useful semantics. This PR causes the transform to remove extra levels of nesting that are left after a Select.

### CODE REVIEW

**Comment by reuvenlax** [comment]
R: @kanterov

**Comment by kanterov** [comment]
I've started looking, the code makes sense, however, I need more time to think about the idea of automatic unnesting. I'm wondering if we can make it less implicit. 

As for me, I would expect the previous behavior, that's how, for instance, Spark data frames work, IIRC.

**Comment by reuvenlax** [comment]
The new behavior better matches how SQL works (Select a.b return an int if
b is an int. It doesn't return a nested row). However we could also make it
an option on the Select transform so the user can pick which behavior they
want.

**Comment by kanterov** [comment]
I checked a similar query in BigQuery:
```
SELECT location FROM UNNEST([
  STRUCT(STRUCT(1.0 as latitude, 2.0 as longtitude) as location, "abc" as userId)
]);
```

It returns:
```
[
  {
    "location": {
      "latitude": "1.0",
      "longtitude": "2.0"
    }
  }
]
```

I understand this PR, it make Beam to return:
```
[{
  "latitude": "1.0",
  "longtitude": "2.0"
}]
```

**Comment by reuvenlax** [comment]
what happens without UNNEST. If you simply do SELECT a.location?

**Comment by reuvenlax** [comment]
Leaving the nested structure in place seem like it would be surprising to
users. For example:

Event:
   location: ROW(Location)

A user would expect to be able to write the following:
  pc.apply(ParDo.of(new DoFn....
      @ProcessElement
      public void process(@FieldAccess("location") Location loc) {
     }));

(where Location is a POJO that matches the Location schema). However this
conversion will fail if the select actually returns an extra level of
nesting. In addition they would expect to be able to write:

  pc.apply(ParDo.of(new DoFn....
      @ProcessElement
      public void process(@FieldAccess("location.lat") double lat) {
     }));

But again, that will fail if the select leaves extra layers of nesting.

**Comment by kanterov** [comment]
As I understand it, `FieldAccess("location")` would work only if we leave an extra layer of nesting. 

I did few BigQuery experiments:
```
CREATE TABLE IF NOT EXISTS test.test (
  userId STRING,
  location STRUCT<latitude FLOAT64, longtitude FLOAT64>
);

INSERT INTO test.test (userId, location) VALUES ("abc", (123.0, 234.0));

SELECT location FROM test.test;
-- Row  location.latitude   location.longtitude 
-- 1    123.0               234.0

SELECT location.latitude FROM gleb_test.test;
-- Row latitude    
-- 1   123.0

```

**Comment by kanterov** [comment]
I need to check how `FieldAccess` works today

**Comment by reuvenlax** [comment]
But doesn't your BigQuery experiment confirm this behavior? You got back a
Row containing location.latitude and location.longtitude, not a Row of Row.
It does look like BigQuery flattened the filed names though.

**Comment by kanterov** [comment]
It created a schema with nested records, to confirm I did an experiment:
```sql
CREATE TABLE `test.test_schema` AS
SELECT location FROM test.test;
```

And I got:

```
location | RECORD | NULLABLE |  
-- | -- | -- | --
location. latitude | FLOAT | NULLABLE |  
location. longtitude | FLOAT | NULLABLE
```

**Comment by kanterov** [comment]
This way it's more visible:
```
$ bq show --schema test.test_schema | jq
[
  {
    "fields": [
      {
        "type": "FLOAT",
        "name": "latitude"
      },
      {
        "type": "FLOAT",
        "name": "longtitude"
      }
    ],
    "type": "RECORD",
    "name": "location"
  }
]```

**Comment by reuvenlax** [comment]
+Kenn for comment

This seems very surprising to me, to be honest. It's not what I would
expect. This behavior has some mathematical niceties (e.g. it makes select
distribute across a union of selectors), but seems like it's almost never
what a user actually wants.

In Beam, part of the goal of schemas is to interact seamlessly with user
types (such as Pojos), so users don't have to deal with Row objects (unless
they want to). The above example for ParDo is one (though the
implementation for that is not yet merged). Another examples is that users
might expect to be able to write:

PCollection<Location> locations = pc.apply(Select.fieldNames("location"))

 .apply(Convert.to(Location.class));

And that will only work if the selected item matches the schema of
Location, not if it's a nested schema.

Maybe for now we should add an option to the Select transform so that the
user can specify which behavior they want?

**Comment by kanterov** [comment]
I did a similar experiment with Spark:
```
df = spark.createDataFrame(
    [{"userId":"abc", "location":{"longtitude":1.0,"latitude":2.0}}], 
    T.StructType([
        T.StructField("userId", T.StringType()),
        T.StructField("location", T.StructType([
            T.StructField("longtitude", T.DoubleType()),
            T.StructField("latitude", T.DoubleType()),
        ])),        
    ])
)


df
> DataFrame[userId: string, location: struct<longtitude:double,latitude:double>]

df.select("location")
> DataFrame[location: struct<longtitude:double,latitude:double>]

df.select("location.latitude")
> DataFrame[latitude: double]
```

**Comment by reuvenlax** [comment]
So select(location.latitude) in Spark matches the unnest semantics, but
select(location) does not. So Spark is inconsistent in itself?

**Comment by kanterov** [comment]
As I understand, the approach is to take the last element in the field path, as I see, BigQuery works the same way:
```
CREATE TABLE `test.test_schema_2` AS
SELECT location.latitude FROM test.test;
```
```
$ bq show --schema test.test_schema_2 | jq
[
  {
    "type": "FLOAT",
    "name": "latitude"
  }
]
```

**Comment by kanterov** [comment]
It seems that Spark allows duplicate column names:

```
df.select("location.latitude", "location.latitude").printSchema()
root
 |-- latitude: double (nullable = true)
 |-- latitude: double (nullable = true)
```
And BigQuery doesn't:
```
Duplicate column names in the result are not supported. Found duplicate(s): latitude
```

**Comment by kennknowles** [comment]
I don't actually see a contradiction here. When you `SELECT location` or when you `df.select("location")` the output is a row with a single column containing a location struct. This is because the value of the selected field is a struct and it gets a default name. That's what this is:

```
[
  {
    "location": {
      "latitude": "1.0",
      "longtitude": "2.0"
    }
  }
]
```

It isn't maintaining the original structure of the input row, it is a list of rows, and each row has a column you selected, and that column contains a struct.

**Comment by kennknowles** [comment]
I'm curious about this one:

```
CREATE TABLE IF NOT EXISTS test.test (
  userId STRING,
  location STRUCT<latitude FLOAT64, longtitude FLOAT64>
);

INSERT INTO test.test (userId, location) VALUES ("abc", (123.0, 234.0));

SELECT location FROM test.test;
-- Row  location.latitude   location.longtitude 
-- 1    123.0               234.0

SELECT location.latitude FROM gleb_test.test;
-- Row latitude    
-- 1   123.0
```

The output when you `SELECT location` is flattened. It should be one column containing a row with the two fields. Is it just for display purposes?

**Comment by reuvenlax** [comment]
Ok this makes sense then.

As I see it BigQuery (and Spark) is unnesting just like this PR doees. The
difference is that it then creates a new Row to contain the selected
result. So select("location") returns a Row containing a location.
select("location.latitude") returns a Row containing a double.  I would
assume that select("location.*") would return a row containing a latitude
and a longitude.

We could validate this by nesting the location three layers deep. In that
case I suspect BQ would return the same thing it did in the above
experiment, not a three-layered nested row.

**Comment by kanterov** [comment]
@kennknowles it isn't flattened, you are right:
```
SELECT location FROM test.test;
```
Gives
```
[
  {
    "fields": [
      {
        "type": "FLOAT",
        "name": "latitude"
      },
      {
        "type": "FLOAT",
        "name": "longtitude"
      }
    ],
    "type": "RECORD",
    "name": "location"
  }
]
```

**Comment by kanterov** [comment]
Makes sense. I tried and both BigQuery and Spark have consistent behavior for selecting rows nested in rows:
```
df.printSchema()
root
 |-- userId: string (nullable = true)
 |-- position: struct (nullable = true)
 |    |-- location: struct (nullable = true)
 |    |    |-- longtitude: double (nullable = true)
 |    |    |-- latitude: double (nullable = true)

df.select("position.location").printSchema()
root
 |-- location: struct (nullable = true)
 |    |-- longtitude: double (nullable = true)
 |    |-- latitude: double (nullable = true)

```

**Comment by reuvenlax** [comment]
Always adding one extra row around whatever is selected makes sense in
Spark and BQ, because it makes selection consistent between primitive and
nested types. Select must always return a row, so if you select userId, it
returns a Row with a single string field (not a scalar string field). Spark
and BQ have therefore decided to treat nested rows the exact same way: even
though the nested row could be returned directly, it is treated
consistently like primitive types and boxed inside a result row.

**Comment by reuvenlax** [comment]
@kanterov I'm changing the semantics here to always return a new row, matching BigQuery and Spark. Will ping once that's done.

**Comment by kanterov** [comment]
@reuvenlax To avoid miscommunication, I didn't reply anything not because I wasn't certain about the solution or something wasn't clear, but because I didn't have time to continue the review.

**Comment by reuvenlax** [comment]
@kanterov I've changed the implementation to always wrap the selected output with a Row. This matches what SQL and Spark does, and is also more self-consistent: it means that selecting a nested Row is the same as selecting a nested Integer).

One concern is that users will want to use the Convert function to convert the result to their POJO (e.g. .apply(Select.fieldNames("location")).apply(Convert.to(Location.class)). So I've also modified the Convert transform to allow automatic unboxing of Rows in this case.

I will have push one more commit adding extra test coverage, however the functionality is already pushed.

**Comment by reuvenlax** [comment]
New unit tests pushed

**Comment by kanterov** [comment]
Sounds good. I will take a look tomorrow

**Comment by kennknowles** [comment]
The Spark behavior looks right to me.

**Comment by reuvenlax** [comment]
@kennknowles I'm not 100% sure whether the better behavior is what Spark is doing or not. @kanterov  do you know what Spark does if you select from an array of an array (without the row in between)?

**Comment by reuvenlax** [comment]
@kennknowles The Spark behavior might be better because it preserves the resulting structure if more field are added. I mean the following. Consider this schema:

S1:
   a: ARRAY[ROW(S2)]
S2: 
  b: INT32
  c: ARRAY[ROW(S3)]
S3:
  d: INT32

now if you Select.fieldNames("a.c.d"), with my current PR you will get back { ARRAY[ARRAY[{d: INT32}]] }. This itself is fine, but consider what happens  if you Select.fieldNames("a.c.d", "a.b") with this PR. You will suddenly get back { ARRAY[{ b: INT32, c: ARRAY[{ d: INT32} ]] }. A new row suddenly got inserted because we needed an extra field. If we adopt the Spark approach, then the structure always stays the same, because the row is already there.

**Comment by reuvenlax** [comment]
@kanterov so far I can't repro this actually. I wrote a test, but it appears to match the Spark behavior. I'll try to exactly copy your example and see what happens

**Comment by reuvenlax** [comment]
Ok, I was reading the Spark result backwards. Spark is eliding the intermediate Row (just returning [][]) and we are not. The behavior in this PR has the advantage of not suddenly changing when two fields are selected, which is why I chose it.

**Comment by kanterov** [comment]
It's hard to read outputs of Spark vs. Beam, I will try to put it in a better way.

Given an example from @reuvenlax:
```
S1:
  a: ARRAY[ROW(S2)]

S2:
  b: INT32
  c: ARRAY[ROW(S3)]

S3:
  d: INT32
``` 

Outputs are:

```
|                      | spark                                                | beam                                                    |
|----------------------|------------------------------------------------------|---------------------------------------------------------|
| select("a.c")        | `struct<c:array<array<struct<d:int>>>>`              | `struct<a:array<struct<c:array<struct<d:int>>>>>`       |
| select("a.b")        | `struct<b:array<int>>`                               | `struct<a:array<struct<b:int>>>`                        |
| select("a.c", "a.b") | `struct<c:array<array<struct<d:int>>>,b:array<int>>` | `struct<a:array<struct<b:int,c:array<struct<d:int>>>>>` |
```

**Comment by kanterov** [comment]
I would try to summarize my understanding of Spark properties:
- resulting schema of `select(<xs : x>)`, would always have a field named `x`, and schema will match the corresponding schema for `<xs : x>` path in the input schema
- if there is `select(<path1>, <path2>)`, resulting schema is union of `select(<path1>)` and `select(<path2>)`

**Comment by kanterov** [comment]
I also uploaded the notebook I used for Spark: https://gist.github.com/kanterov/78972d40ebe5b6f5c553b5a018dbb0bf

You can get local Spark instance with Jupyter with:
```
$ docker run -p8888:8888 -it jupyter/pyspark-notebook
```

Or using Dataproc

**Comment by kanterov** [comment]
Spark has an implementation of schema prunning, that behaves similar to the code in this PR (preserves schema structure). It was introduced in https://github.com/apache/spark/pull/21320, there is a test suite demonstrating it's behavior https://github.com/apache/spark/pull/21320/files#diff-3131013c95ca682e798070a2d50d6896. 

As I understand, it isn't used in user-facing API, only for pushing down projections into Parquet data source.

**Comment by kennknowles** [comment]
Oh, I see now. The issue is eliding the array layers actually.

**Comment by kennknowles** [comment]
I think the summary by @kanterov is super clear are the right approach, but I don't know if I would call that "union" so much as row building.

For SQL I would state this as `SELECT <expr1 : name1>, <expr2, name2> ...` each result is a new row containing the columns `name1`, `name2` etc and the type of each column is the type of the expression.

Now for raw Beam if you select just a single column and the type of that column can convert to a Java type, you shoulud be able to convert. Specifically:

 - row with one int column convertible to int
 - row with one row column matching a POJO can convert directly

This is a step towards a thing mentioned in the Go SDK coders thread:

 - row with one bytes column w/ coder metadata equivalent to today's approach of coders

**Comment by reuvenlax** [comment]
Given that this is mostly an issue eliding nested arrays, I wonder if we
should merge this PR and file a JIRA for the nested array case. We can
discuss the correct behavior on the JIRA. This PR seems more correct than
the old behavior.

**Comment by reuvenlax** [comment]
@kanterov I'm surprised that Spark cannot resolve select("a.c.d"). Is there a reason for that?

**Comment by kanterov** [comment]
@reuvenlax I was surprised to see that Spark didn't resolve `select("a.c.d")`

Agree that we should keep iterations quick, and given API is heavily experimental we could move forward with increments. I think we did a good analysis, and now the problem better. However, we still don't have a good formalization of desired semantics.

I'm going to revisit the code again, didn't have time to check this today.

**Comment by kanterov** [comment]
@reuvenlax I've tried the following test:
```
  @Test
  public void testSelectFieldOfRecord() {
    Schema f1 = Schema.builder().addInt64Field("f0").build();
    Schema f2 = Schema.builder().addRowField("f1", f1).build();
    Schema f3 = Schema.builder().addRowField("f2", f2).build();
    
    Row r1 = Row.withSchema(f1).addValue(42L).build(); // {"f0": 42}
    Row r2 = Row.withSchema(f2).addValue(r1).build();  // {"f1": {"f0": 42}}
    Row r3 = Row.withSchema(f3).addValue(r2).build();  // {"f2": {"f1": {"f0": 42}}}

    FieldAccessDescriptor fieldAccessDescriptor = FieldAccessDescriptor
        .withFieldNames("f2.f1")
        .resolve(f3);

    Schema outputSchema = SelectHelpers.getOutputSchema(f3, fieldAccessDescriptor);

    Row out = SelectHelpers.selectRow(r3, fieldAccessDescriptor, r3.getSchema(), outputSchema);

    assertEquals(outputSchema, f2);
    assertEquals(out, r2);
  }
```

And it fails with an exception:

```
ava.lang.IllegalArgumentException: For field name f0 and type INT64 found incorrect class type class org.apache.beam.sdk.values.RowWithStorage
	at org.apache.beam.sdk.values.Row$Builder.verifyPrimitiveType(Row.java:724)
	at org.apache.beam.sdk.values.Row$Builder.verify(Row.java:587)
	at org.apache.beam.sdk.values.Row$Builder.verify(Row.java:571)
	at org.apache.beam.sdk.values.Row$Builder.build(Row.java:748)
	at org.apache.beam.sdk.schemas.utils.SelectHelpers.selectIntoRow(SelectHelpers.java:204)
	at org.apache.beam.sdk.schemas.utils.SelectHelpers.selectRow(SelectHelpers.java:167)
	
```

I don't see anything wrong with the test itself.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/transforms/Convert.java [hunk 1]
```diff
@@ -92,11 +99,24 @@
...
     ConvertTransform(TypeDescriptor<OutputT> outputTypeDescriptor) {
       this.outputTypeDescriptor = outputTypeDescriptor;
     }
 
+    @Nullable
+    private static Schema getBoxedNestedSchema(Schema schema) {
+      if (schema.getFieldCount() > 1) {
+        return null;
+      }
+      FieldType fieldType = schema.getField(0).getType();
```
[the remark this answers is not in the document]
Comment (reuvenlax) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/transforms/Convert.java]: Right now the entire Convert transform ignores nullable (this was already a bit of a bug with it). We appear to be missing a JIRA on this issue, so I'll file one.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/utils/SelectHelpers.java [hunk 2]
```diff
@@ -106,47 +164,89 @@ public static Row selectRow(
...
+    selectIntoRow(input, output, fieldAccessDescriptor, inputSchema);
+    return output.build();
+  }
+
+  /** Select out of a given {@link Row} object. */
+  public static void selectIntoRow(
+      Row input,
+      Row.Builder output,
+      FieldAccessDescriptor fieldAccessDescriptor,
+      Schema inputSchema) {
```
[the remark this answers is not in the document]
Comment (reuvenlax) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/utils/SelectHelpers.java]: Yes. removed

File: sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/transforms/Convert.java [hunk 3]
```diff
@@ -92,11 +99,24 @@
...
     TypeDescriptor<OutputT> outputTypeDescriptor;
+    Schema unboxedSchema = null;
 
     ConvertTransform(TypeDescriptor<OutputT> outputTypeDescriptor) {
       this.outputTypeDescriptor = outputTypeDescriptor;
     }
 
+    @Nullable
+    private static Schema getBoxedNestedSchema(Schema schema) {
+      if (schema.getFieldCount() > 1) {
```
[the remark this answers is not in the document]
Comment (reuvenlax) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/transforms/Convert.java]: Done

**Comment by reuvenlax** [comment]
Good catch on that test. This was a bug in my code. I've fixed that bug and added your test.

**Comment by reuvenlax** [comment]
@kennknowles and I discussed the array issue. We both think that the best approach is probably to distribute out array selects. However there are other options that should be discussed (e.g. completely unnesting arrays whenever a select happens), so we think this discussion should be brought to the dev list. This PR I believe is generally an improvement, so (assuming other comments are resolved) I think we should merge it and address array select semantics separately.

**Comment by reuvenlax** [comment]
FYI by "distribute array selects" I mean the following:

if you have { a: {b: int, c: int}[] } and someone select "a.b", "a.c" the result should be {b: int[], c: int[]}. The same should be true if someone selects a.*. However if someone selects just a, you get the original schema back.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/schemas/utils/SelectHelpersTest.java [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+
+    FieldAccessDescriptor fieldAccessDescriptor =
+        FieldAccessDescriptor.withFieldNames("f2.f1").resolve(f3);
+
+    Schema outputSchema = SelectHelpers.getOutputSchema(f3, fieldAccessDescriptor);
+
+    Row out = SelectHelpers.selectRow(r3, fieldAccessDescriptor, r3.getSchema(), outputSchema);
+
+    assertEquals(outputSchema, f2);
+    assertEquals(out, r2);
```
Comment (kanterov) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/schemas/utils/SelectHelpersTest.java]: Parameters should be in a different order: expected, actual

**Comment by kanterov** [comment]
I think we can merge and iterate from there, but there are still issues with semantics that can be discussed separately.

**Comment by reuvenlax** [comment]
@kanterov looking. I suspect my fix to the last issue you found is what caused this use case to break.

**Comment by reuvenlax** [comment]
@kanterov I believe actually that the bug is in your test.  You have:
    Schema outputSchema = SelectHelpers.getOutputSchema(f3, fieldAccessDescriptor);
when it should be
    Schema outputSchema = SelectHelpers.getOutputSchema(f4, fieldAccessDescriptor);

When I fix this, the test now passes.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/schemas/utils/SelectHelpersTest.java (hunk 4 above)
Comment (reuvenlax) [reply sdks/java/core/src/test/java/org/apache/beam/sdk/schemas/utils/SelectHelpersTest.java]: done

**Comment by reuvenlax** [comment]
Run Java PreCommit

**Comment by reuvenlax** [comment]
Run Java PreCommit

**Comment by reuvenlax** [comment]
Run Java PreCommit

**Comment by reuvenlax** [comment]
@kanterov I went ahead and changed array and map selects to distribute the select. This preserves the invariant that the field name selected is always the one that appears in the resulting schema, and I believe is closer to what Spark does. 

I still think we should have a thread on the dev list to discuss the detailed semantics, as we still might want to change them. However I think with this last commit, we are probably closer to where we want to end up.

**Comment by reuvenlax** [comment]
@kanterov did you have any further comments on this PR?

**Comment by kanterov** [comment]
@reuvenlax sorry for the delay, LGTM 👍 please feel free to merge when you are comfortable with it

**Pull request merged**